test: Update e2e docker template

mattrunyon opened the pull request:
While helping Ethan debug e2e flaky tests in CI, Don mentioned the npm script to run e2e in docker was ugly. I agreed and decided to clean it up some.

This change makes the e2e tests in Docker also spin up the DHC server w/ anonymous auth so you don't have to. We could extend this to add a dev mode for e2e testing possibly, but I don't think you can get headed mode in Docker if you want it.

Also figured out a way to remove a COPY line for every package.json using multi-stage builds. It caches properly still if no package.json files change.

Removed the volume mount for test-results because Docker creates the directories inside of it as root (I tried running the container as a regular user and it complains). This causes an issue if you run the snapshots in Docker w/ a failure and then try to run locally because the folder is owned by root and can't be cleared as a regular user.

Still need to update the readme.

mattrunyon commented on tests/docker-scripts/docker-compose.yml:
```diff
@@ -0,0 +1,34 @@
+
+services:
+  dhc-server:
+    image: ghcr.io/deephaven/server:${VERSION:-latest}
```
Good catch. I agree it doesn't make too much sense to test with locally built versions because that test would fail in CI if the core change isn't merged first

mattrunyon commented:
@mofojed In the last commit I also moved the `webServer` part of the playwright config to the CI config only (also used in the docker images). I found the old method to be an easy way to accidentally test against a stale local build if you forgot to start the dev server. You can still test against a local build by just running the app preview script

mofojed commented on README.md:
```diff
@@ -88,13 +80,47 @@ A standalone application with it's own entry point. Recommend copying the [embed
+
+Snapshots are used by end-to-end tests to visually verify the output. Snapshots are both OS and broser specific. Sometimes changes are made requiring snapshots to be updated. Update snapshots locally by running `npm run e2e:update-snapshots`.
+
+Once you are satisfied with the snapshots and everything is passing locally, you need to use the docker image to update snapshots for CI (unless you are running the same platform as CI (Ubuntu)). Run `npm run e2e:update-ci-snapshots` to update the CI snapshots. The snapshots will be written to your local directories and should be committed to git.
```
Might want to mention somewhere in here that the local ones are ignored (e.g. snapshots for the darwin platform)

mofojed commented on README.md:
```diff
@@ -88,13 +80,47 @@ A standalone application with it's own entry point. Recommend copying the [embed
+
+- `npm run e2e`: Run end-to-end tests in headless mode.
+- `npm run e2e:headed`: Runs end-to-end tests in headed debug mode. Also ignores snapshots since a test suite will stop once 1 snapshot comparison fails. Useful if you need to debug why a particular test isn't working. For example, to debug the `table.spec.ts` test directly, you could run `npm run e2e:headed -- ./tests/table.spec.ts`.
+- `npm run e2e:codegen`: Runs plawright in codegen mode which can help with creating tests. See [Playwright Codegen](https://playwright.dev/docs/codegen/) for more details.
```
Typo - `playwright`, not `plawright` (not sure why I can't suggest the fix)

mofojed reviewed (changes requested):
Looks good, just one minor typo

mofojed reviewed (changes requested):
I made a bunch of changes to IrisGrid to make it look completely different, then tried running this on my mac:
```
npm install
npm run e2e:update-ci-snapshots
```

It doesn't look like any snapshots were actually updated though.

mattrunyon commented:
Digging into the issue, I think the dockerignore is not ignoring the dist folders and #1339 means we're consuming whatever the locally built files are. I ran a clean, then `npm ci` then `npm run build:app` locally and got an error about our packages not resolving to anything. I'll fix the dockerignore and switch to build the packages first, then the app (or just do a full build)

mattrunyon commented on .dockerignore:
```diff
@@ -1,22 +1,17 @@
 
 # dependencies
-/node_modules
+node_modules*
```
I actually have this and the build/dist wrong. Docker's ignore is not like `.gitignore` and doesn't match these at any depth like git does. Instead they should be `**/node_modules/` I think

mattrunyon commented:
Ok I think the dockerignore file is good now. Tested running `npm run clean` (we clean nx cache but don't install nx currently which errors before finishing the clean) and then `npm run e2e:docker` (cached build properly).

Ran `npm run build` then `npm run e2e:docker` and it was still cached

Changed a src file and ran `npm run e2e:docker` and it triggered a rebuild

Changed a `package.json` and ran `npm run e2e:docker` and it triggered a reinstall and rebuild
